DIPs/1NNN-WGB.md

```diff
@@ -0,0 +1,155 @@
+# Argument Ownership and Function Calls
+
+
+| Field           | Value                                                           |
+|-----------------|-----------------------------------------------------------------|
+| DIP:            | XXXX                                                            |
+| Review Count:   | 0                                                               |
+| Author:         | Walter Bright                                                   |
+| Implementation: |                                                                 |
+| Status:         |                                                                 |
+
+
+## Abstract
+
+DIP25 and DIP1000 greatly improved the memory safety of passing
+references and pointers to functions by checking whether those
+pointers escaped a function or not. This implies that a container
+can safely pass a reference to its internals to a function, if that
+function doesn't escape it.
+
+But if the function is also passed another reference to the container,
+that reference can render invalid what the other reference refers to.
+This DIP aims to correct that problem. It's a natural progression after
+DIP25 and DIP1000.
+
+## Contents
+* [Rationale](#rationale)
+* [Description](#description)
+* [Breaking Changes and Deprecations](#breaking-changes-and-deprecations)
+* [Reference](#reference)
+* [Copyright & License](#copyright--license)
+* [Reviews](#reviews)
+
+## Rationale
+
+Containers cannot be memory safe if they cannot control memory safe
+access to their payloads. Containers cannot be efficient if they cannot
+make available direct references to their payloads. Entreating the user
+to just not do certain things is not reliable and does not scale.
+
+The simplest illustration of the problem:
+
+```
+struct S {
+    byte* ptr;
+    ref byte get() { return *ptr; }
+}
+
+void foo(ref S t, ref byte b) {
+    free(t.ptr);    // frees memory referred to by b
+    b = 4;          // corrupt memory access
+}
+
+void test() {
+    S s;
+    s.ptr = cast(byte*)malloc(1);
+    foo(s, s.get());  // (*)
+}
+```
+The same problem using scope pointers:
+```
+struct S {
+    byte* ptr;
+    byte* byte get() { return ptr; }
+}
+
+void foo(scope S* t, scope byte* pb) {
+    free(t.ptr);    // frees memory referred to by pb
+    *pb = 4;        // corrupt memory access
+}
+
+void test() {
+    S s;
+    s.ptr = cast(byte*)malloc(1);
+    foo(&s, s.get());  // (*)
+}
+```
+
+D currently has no defense against this problem. It makes checkably memory safe
+reference counting not possible. (Timon Gehr first pointed this out.)
+
+
+## Description
+
+The solution hinges on the recognition that, in the example, two mutable references
+to the same data are passed to the function `foo()`. Generalizing, whenever there are
+multiple references to the same data, and one of them is mutable, the mutable reference
+can be used to invalidate the data the other references refer to, whether they are
```

Comment:
as above, please label references for clarity

Reply:
I don't think that'll help here.